3D mortar frictional contact

recuero opened the pull request:
This is a WIP PR ultimately intended to create consistent 3D frictional contact forces.

Some items

- Enable nodal quantities for the description of geometry (orthonormal normal-tangents vectors)
- Extend `ComputeFrictionalForceLMMechanicalContact` to also perform 3D frictional constraint enforcement
- Allow, if possible, displacement residual objects to act only on surface nodes

Convergence with increased deformation isn't optimal and may lead to some models not converging at all. Possible solutions include improvements to the Jacobian and adding an augmented Lagrange loop. Or better yet, all of the above.

(#17495)

recuero committed:
Fix uncaught regression on the use of nodal interpolated normals (#17495)

When Alex implemented the possibility of using nodal normals to enforce the
normal contact constraints, I failed to add corresponding tests and later
changes caused regression of this capability, which I plan on using to
create a consistent tangent vector field for three dimensional frictional
constraints.

recuero committed:
Initial take on 3D frictional mortar constraints (#17495)

Not functioning, it remains to:

 - Create consistent tangential vector field in mortar segment generation classes
 - Use nodal based interpolated, consistent tangents
 - Have some initial verification

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/899430/) on 656e32c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19462/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19462/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f2f4113880736a8ce6dc2289d54efc4bf01a74a3`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/902184/) on 8b0b30f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19462/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19462/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 36b3e788ae186e0923236f3bb0b3814b648f9cd4`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/902217/) on c768a90 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19462/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19462/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 36b3e788ae186e0923236f3bb0b3814b648f9cd4`

jbadger95 commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -65,6 +71,11 @@ ComputeFrictionalForceLMMechanicalContact::ComputeFrictionalForceLMMechanicalCon
+  if (_3d && !_friction_var_dir)
+    paramError("friction_lm_dir",
+               "Three-dimensional mortar frictional contact simulations require an additional "
+               "frictional Lagrange's multiplier to enforce a second tangential pressure");
```
Cool to see this getting fixed up! I'm excited to see this, I'll take a closer look when you're ready, just wanted to point out that below I neglected to compute the z component of the relative velocity, that should probably be added to the `if (_has_disp_z)` part below where the derivative is being calculated.

recuero replied:
It should be fixed now.

recuero committed:
Build maps from interior parents to lower-dimensional dimensional domains.

Also:

 - Create tangent vector fields that have a chance to work
 - Miscellaneous changes and fixes

recuero committed:
Add ADMortarLagrangeConstraint that skips interior parent nodes

Make according changes to contact classes to inherit from
ADMortarLagrangeConstraint so that the displacement residual
from the contact constraints can leverage contact geometry
particularized at the nodes.

Right now, this is enabled through the flag interpolate_normals.
We should eventually change this to something like use_nodal_geometry.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/903941/) on a0819a1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19462/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19462/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 97d5b7b17e145d7aa912802bff51f34763dda2df`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/904207/) on 1051b8f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19462/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19462/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 97d5b7b17e145d7aa912802bff51f34763dda2df`

recuero committed:
Initial woking implementation  (#17495)

- Settled on householder algorithm to obtain orthonormal basis (tangent vectors)
- Sphere on plane with friction runs and appears to obtain physical results
- Convergence is poor, definitely needs improvement
- Need to add tests

recuero committed:
Clean up and address Alex's comments

- Modify frictional contact constraint data structures to avoid code duplication
- Fix merge in ComputeMortarFunctor after 3D dual mortar changes.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/905241/) on ef3a968 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19462/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/905251/) on ef3a968 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>5a2f36</th>
      <th colspan=4>#19462 ef3a96</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.07%</td>
      <td>82.02%</td>
      <td>-0.05%</td>
      <td>52.55%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72516</td>
      <td>72586</td>
      <td>+70</td>
      <td>72</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15846</td>
      <td>15911</td>
      <td>+65</td>
      <td>65</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19462/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19462/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>5a2f36</th>
      <th colspan=4>#19462 ef3a96</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>87.90%</td>
      <td>85.54%</td>
      <td>-2.36%</td>
      <td>43.41%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2273</td>
      <td>2301</td>
      <td>+28</td>
      <td>56</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>313</td>
      <td>389</td>
      <td>+76</td>
      <td>73</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19462/coverage_diff/contact.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19462/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19462/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 52.55% is less than the suggested 90.0%
- `contact` new line coverage rate 43.41% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/905251/) on ef3a968 wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 85.54% is less than the required 87.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/905916/) on 64bcf5f wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 85.54% is less than the required 87.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/906310/) on 9e59d8e wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 85.54% is less than the required 87.0%

lindsayad commented on modules/contact/include/constraints/ComputeFrictionalForceLMMechanicalContact.h:
```diff
@@ -57,12 +65,29 @@ class ComputeFrictionalForceLMMechanicalContact : public ComputeWeightedGapLMMec
   /// A map from node to weighted gap
   std::unordered_map<const DofObject *, ADReal> _dof_to_weighted_tangential_velocity;
```
Is the doc string a typo, or the actual data member a typo?

lindsayad commented on modules/contact/include/constraints/ComputeFrictionalForceLMMechanicalContact.h:
```diff
@@ -57,12 +65,29 @@ class ComputeFrictionalForceLMMechanicalContact : public ComputeWeightedGapLMMec
   /// A map from node to weighted gap
   std::unordered_map<const DofObject *, ADReal> _dof_to_weighted_tangential_velocity;
 
+  /// A map from node to weighted gap for an additional tangential direction (three dimensions)
+  std::unordered_map<const DofObject *, ADReal> _dof_to_weighted_tangential_velocity_dir;
```
same question here. You're talking about gaps in the doc string but the data member is tangential velocity. Also does it make sense to have the original data member be a `std::unordered_map<const DofObject *, std::vector<ADReal>` and not have this additional new data member?

recuero replied:
Doc strings need to be revised throughout. Yeah, will update the data member to avoid new member.

lindsayad commented on modules/contact/include/constraints/ComputeFrictionalForceLMMechanicalContact.h:
```diff
@@ -57,12 +65,29 @@ class ComputeFrictionalForceLMMechanicalContact : public ComputeWeightedGapLMMec
   /// The value of the tangential velocity at the current quadrature point
   ADReal _qp_tangential_velocity;
 
+  /// The value of the tangential velocity at the current quadrature point
+  /// in an additional direction, required for three-dimensional cases
+  ADReal _qp_tangential_velocity_dir;
+
+  /// The value of the tangential velocity at the current node
+  ADRealVectorValue _qp_tangential_velocity_nodal;
+
+  /// The value of the tangential velocity at the current node
+  /// in an additional direction, required for three-dimensional cases
+  ADRealVectorValue _qp_tangential_velocity_dir_nodal;
```
Begs the same question here. A `_dir` suffix doesn't really communicate to me what the data member is for, e.g. that it's for holding an additional direction when in 3D vs. 2D. I would vote to make the original data members `std::vectors` and size them in the constructor according to the mesh dimension

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -100,7 +127,27 @@ ComputeFrictionalForceLMMechanicalContact::computeQpIProperties()
+  // Get the _dof_to_weighted_tangential_velocity map for a second direction
+  if (_3d)
+  {
+    const DofObject * dof_dir =
```
```suggestion
    const DofObject * const dof_dir =
```

lindsayad reviewed (changes requested):
I did not do a thorough review but I just wonder if we generalize our current data and methods and use loops over number of tangential directions whether we can avoid adding so many new lines. I'm wondering whether it's possible to avoid any of these new `_dir` members

lindsayad commented on framework/src/constraints/ADMortarLagrangeConstraint.C:
```diff
@@ -0,0 +1,177 @@
+  residuals.resize(test_space_size, 0);
+  for (_qp = 0; _qp < _qrule_msm->n_points(); _qp++)
+    for (_i = 0; _i < test_space_size; _i++)
+      residuals[_i] += _JxW_msm[_qp] * _coord[_qp] * computeQpResidual(mortar_type);
```
I think we would like to make life in user code as easy as possible. For instance I see in `NormalMortarMechanicalContact` that you have high-level 0 residual that you return if `_i` is not on the face. I would instead propose a design in which you only call `computeQpResidual` when `_i` is on the face. So I would move this code after your dof filtering, and in your dof filtering I would build up something like a `std::vector<unsigned int> is` which only contains the `_i`s that are on the face. Then you can do
```c++
  for (const auto index : is)
  {
    _i = index;
    // call computeQpResidual and stuff
  }
```
I recognize that this then will require you to implement `computeResidual` as well as `computeJacobian` in this class

lindsayad replied:
Then user code should essentially never be returning 0 from `computeQpResidual`

recuero replied:
Can you please inspect the code to see if that's what you described.

recuero replied:
Never mind, after latest commit going back to:

```Assertion `compare_dofs_set == current_dofs_set' failed
We're going to see whether the dof sets are the same. IIRC the degree of freedom dependence (as indicated by the dof index set held by the ADReal) has to be the same for every residual passed to this method otherwise constrain_element_matrix will not work.
at /Users/marta3/projects3/moose/framework/build/header_symlinks/Assembly.h, line 2798```

lindsayad commented on modules/contact/include/constraints/ComputeFrictionalForceLMMechanicalContact.h:
```diff
@@ -49,26 +49,40 @@ class ComputeFrictionalForceLMMechanicalContact : public ComputeWeightedGapLMMec
    */
   virtual void enforceConstraintOnDof(const DofObject * const dof) override;
 
+  /**
+   * Method called from \p post(). Used to enforce node-associated constraints. E.g. for the base \p
+   * ComputeFrictionalForceLMMechanicalContact we enforce PDASS frictional constraints. This is also
+   * where we actually feed the node-based constraint information into the system residual and
+   * Jacobian. Method used to enforce frictional constraints in three-dimensional simulations.
+   */
+  virtual void enforceConstraintOnDof3d(const DofObject * const dof);
```
I was hoping that this method would have disappeared

recuero replied:
The equations for 2d and 3d are quite different, I'd say completely different. I could still merge those two methods if you are so inclined.

lindsayad replied:
Ok then let's not merge them. But per my comment in the .C file, sans the ternary to determine the number of tangents (since we're keeping this method, we know it will be 2), let's make things a loop if we can such that we can get rid of all the `_dir`s

lindsayad commented on modules/contact/include/constraints/ComputeFrictionalForceLMMechanicalContact.h:
```diff
@@ -49,26 +49,40 @@ class ComputeFrictionalForceLMMechanicalContact : public ComputeWeightedGapLMMec
   /// Numerical factor used in the tangential constraints for convergence purposes
   const Real _c_t;
 
   /// Frictional Lagrange's multiplier variable pointer
   MooseVariable * _friction_var;
 
+  /// Frictional Lagrange's multiplier variable pointer for additional 3D direction
+  MooseVariable * _friction_var_dir;
```
```suggestion
  /// Frictional Lagrange's multiplier variable pointer
  std::vector<MooseVariable *> _friction_vars;
```

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -73,20 +84,37 @@ ComputeFrictionalForceLMMechanicalContact::computeQpProperties()
-  // Get the component in the tangential direction
-  // TODO: Create a consistent mortar tangential vector field
-  _qp_tangential_velocity = relative_velocity * (_tangents[_qp][0] * _JxW_msm[_qp] * _coord[_qp]);
+  if (_interpolate_normals)
+    _qp_tangential_velocity[0] =
+        relative_velocity * (_tangents[_qp][0] * _JxW_msm[_qp] * _coord[_qp]);
+  else
+    _qp_tangential_velocity_nodal = relative_velocity * (_JxW_msm[_qp] * _coord[_qp]);
+
+  // If we don't use interpolated normals, we get the directions at the node,
+  // not here at the quadrature points.
+  if (_3d && _interpolate_normals)
+    _qp_tangential_velocity[1] =
+        relative_velocity * (_tangents[_qp][1] * _JxW_msm[_qp] * _coord[_qp]);
```
```suggestion
  if (_interpolate_normals)
  {
    _qp_tangential_velocity[0] =
        relative_velocity * (_tangents[_qp][0] * _JxW_msm[_qp] * _coord[_qp]);
    if (_3d)
      _qp_tangential_velocity[1] =
        relative_velocity * (_tangents[_qp][1] * _JxW_msm[_qp] * _coord[_qp]);
  }
  else
    _qp_tangential_velocity_nodal = relative_velocity * (_JxW_msm[_qp] * _coord[_qp]);
```

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -190,9 +239,95 @@ ComputeFrictionalForceLMMechanicalContact::incorrectEdgeDroppingPost(
+  ADReal contact_pressure = (*_sys.currentSolution())(normal_dof_index);
+  Moose::derivInsert(contact_pressure.derivatives(), normal_dof_index, 1.);
+
+  // Get friction LM
```
I think we can do all of this in `enforceConstraintOnDof` and not introduce `enforceConstraintOnDof3d`. We can do:
```c++
const unsigned int num_tangents = _3d ? 2 : 1;
for (const auto i : make_range(num_tangents))
{
  // Do all the stuff, indexing into our arrays/vectors with i to get the right component
}
```
Then hopefully we can really get rid of all the `_dir` members.

lindsayad reviewed (changes requested):
Still nothing is jumping out at me here as far as concerns regarding Jacobian/parallel consistency (I'm glad you got that latter one ironed out to avoid TIMPI compile issues).

But the nice thing, if we get to a place where we are basically reusing all the same code between 2D and 3D, would be that we get additional coverage of any new code you are adding in our existing 2D test cases.

recuero commented on framework/include/constraints/MortarConstraintBase.h:
```diff
@@ -145,9 +170,18 @@ class MortarConstraintBase : public Constraint,
+  /// Tangent vectors on the secondary faces (libmesh)
   const MooseArray<std::vector<Point>> & _tangents;
 
+  /// Tangents vector on the secondary faces (householder from normal vectors)
+  std::array<std::vector<Point>, 2> _tangents_3d;
```
Yes, they are supposed to. They are obtained with the interpolated but nodal normal vector that is computed in the mortar generation.

recuero committed:
Update modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/912256/) on adbf672 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19462/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19462/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 95408a4979e4e3fe1498261e5504c9b8dba6ed82`

recuero committed:
Changes requested by Alex

 - Clean up additional lines of code due to 3D friction
 - Make ADMortarLagrangeConstraint more user friendly. The user code
   does not need to return zero when node isn't on the surface/face.

recuero reviewed (commented):
Do you know what changes are required to make this method work? I got seg faulted.

https://github.com/idaholab/moose/blob/adbf6727ba86da8c62f9197e0243876c73e396ea/framework/src/constraints/AutomaticMortarGeneration.C#L2016

lindsayad committed:
enable outputting of nodal normals

recuero committed:
Update tests, double braces, testing for tangent geometry output

recuero committed:
Add 3D frictional mortar contact to contact action (#17495)

recuero committed:
Move ADMortarLagrangeConstraint from the framework to the contact module.

Despite this class being general enough to apply nodal residuals, its use
is for now limited to contact applications.

Refs. #17495.

recuero committed:
Add auxiliary kernel to compute mortar tangential frictional pressure vector (#17495)

This nodal kernel is used to compute the frictional pressure vector when using mortar
contact in three-dimensions. The direct output of tangent Lagrange's multipliers is associated
with unit vectors whose orientation is unknown. For this reason, this processing of Lagrange's
multipliers and tangent vectors enables improved result analyses.

lindsayad commented on modules/contact/src/constraints/NormalMortarMechanicalContact.C:
```diff
@@ -44,12 +44,29 @@ NormalMortarMechanicalContact::computeQpResidual(Moose::MortarType type)
+        return _test_secondary[_i][_qp] * _lambda[_qp] * _normals[_qp](_component);
+      else
+      {
+        unsigned int normal_index = 0;
+        normal_index = _secondary_ip_lowerd_map.at(_i);
```
```suggestion
        const auto normal_index = libmesh_map_find(_secondary_ip_lowerd_map, _i);
```
and everywhere else similarly

lindsayad commented on modules/contact/include/constraints/ADMortarLagrangeConstraint.h:
```diff
@@ -0,0 +1,33 @@
+
+#include "ADMortarConstraint.h"
+
+class ADMortarLagrangeConstraint : public ADMortarConstraint
```
Need class doxygen

lindsayad commented on framework/include/constraints/MortarConstraintBase.h:
```diff
@@ -145,9 +170,18 @@ class MortarConstraintBase : public Constraint,
+  /// Tangents vector on the secondary faces (householder from normal vectors)
+  std::array<std::vector<Point>, 2> _tangents_3d;
+
+  /// Map to locate nodes from higher dimensional element to lower dimensional elements
+  std::map<unsigned int, unsigned int> _secondary_ip_lowerd_map;
+
+  /// Map to locate nodes from higher dimensional element to lower dimensional elements
+  std::map<unsigned int, unsigned int> _primary_ip_lowerd_map;
```
Ordinarily I would suggest `std::unordered_map` but these containers should be really small and a binary search in a `std::map` may be faster

lindsayad commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -178,6 +177,26 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
+   */
+  std::array<std::vector<Point>, 2> getNodalTangents(const Elem & secondary_elem) const;
+
+  /**
+   * @return The mapping from secondary interior parent nodes to lower dimensional nodes
+   */
+  std::map<unsigned int, unsigned int>
+  getSecondaryIpToLowerElementMap(const Elem & _lower_secondary_elem) const;
+
+  /**
+   * @return The mapping from primary interior parent nodes to its corresponding lower dimensional
+   * nodes
+   */
+  std::map<unsigned int, unsigned int>
+  getPrimaryIpToLowerElementMap(const Elem & _primary_elem,
+                                const Elem & _primary_elem_ip,
+                                const Elem & _lower_secondary_elem) const;
```
For these methods I might mention in the doxygen string that these maps are built on-the-fly because every time I see them I wonder why we are returning by value and not by reference

lindsayad commented on framework/include/constraints/MortarConstraintBase.h:
```diff
@@ -85,6 +85,31 @@ class MortarConstraintBase : public Constraint,
    */
   void setNormals(const std::vector<Point> & normals) { _normals = normals; }
 
+  /**
+   * Set the nodal tangents
+   */
+  void setNodalTangents(const std::array<std::vector<Point>, 2> & tangents)
+  {
+    _tangents_3d = tangents;
+  }
```
I don't know what to make of the doxygen and method name seemingly being generic and dimension agnostic, but the data member name specifying 3d. I feel like something should change here but I don't know what. Should this method only be called in 3d and hence perhaps the method name should be changed? Or should the data member name be changed?

recuero replied:
These are nodal tangents obtained from an orthogonalization process starting with the interpolated nodal normal vectors. I think I added 3d to make sure I was using those for 3d applications, since it guarantees that they are geometrically and 'mortarly' consistent. But I think we can call them `_nodal_tangents` which will distinguish them from the `tangents` coming from `libmesh`.

lindsayad replied:
I would like that

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1406,6 +1471,13 @@ AutomaticMortarGeneration::computeNodalNormals()
+    // Map to get lower dimensional element from interior parent on secondary surface
+    // This map can be used to provide a handle to methods in this class that need to
+    // operate on lower dimensional elements.
+
+    _secondary_element_to_secondary_lowerd_element.insert(
+        std::make_pair(interior_parent->id(), secondary_elem));
```
```suggestion
    _secondary_element_to_secondary_lowerd_element.emplace(
        interior_parent->id(), secondary_elem);
```

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1933,24 +2034,47 @@ AutomaticMortarGeneration::projectPrimaryNodesSinglePair(
+  if (!_nodal_normals_system)
+  {
+    _nodal_normals_system = &nodal_normals_es.template add_system<ExplicitSystem>("nodal_normals");
+    _nnx_var_num = _nodal_normals_system->add_variable("nodal_normal_x", FEType(FIRST, LAGRANGE)),
+    _nny_var_num = _nodal_normals_system->add_variable("nodal_normal_y", FEType(FIRST, LAGRANGE));
+    _nnz_var_num = _nodal_normals_system->add_variable("nodal_normal_z", FEType(FIRST, LAGRANGE));
+
+    _t1x_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_1_x", FEType(FIRST, LAGRANGE)),
+    _t1y_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_1_y", FEType(FIRST, LAGRANGE));
+    _t1z_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_1_z", FEType(FIRST, LAGRANGE));
+
+    _t2x_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_2_x", FEType(FIRST, LAGRANGE)),
+    _t2y_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_2_y", FEType(FIRST, LAGRANGE));
+    _t2z_var_num =
+        _nodal_normals_system->add_variable("nodal_tangent_2_z", FEType(FIRST, LAGRANGE));
+    nodal_normals_es.reinit();
```
I know I wrote this, but would we ever want to output the non-vertex normals?

recuero replied:
The result of the normal vector interpolation? Maybe, but I think outputting the normals at the vertices already provides good information.

lindsayad replied:
No these ones aren't actually interpolated IIRC. We have concrete nodal normals at non-vertex nodes I'm pretty sure... my memory is so bad. yea ok it looks like I'm right
```c++
std::vector<Point>
AutomaticMortarGeneration::getNodalNormals(const Elem & secondary_elem) const
{
  std::vector<Point> nodal_normals(secondary_elem.n_nodes());
  for (const auto n : make_range(secondary_elem.n_nodes()))
    nodal_normals[n] = _secondary_node_to_nodal_normal.at(secondary_elem.node_ptr(n));

  return nodal_normals;
}
```
We loop over all the nodes here, not just vertices

recuero replied:
I see, didn't understand at first. We would need to make quite a few changes to this method to handle more general cases, e.g. select the interpolation order for output. But, yeah, you could see how potentially we would want to check geometry information for non-vertex nodes when applicable.

lindsayad replied:
can do in the future as desired

lindsayad commented on framework/src/loops/ComputeMortarFunctor.C:
```diff
@@ -267,6 +272,9 @@ ComputeMortarFunctor::operator()()
         for (auto * const mc : _mortar_constraints)
         {
           mc->setNormals(mc->interpolateNormals() ? normals : nodal_normals);
+          mc->setSecondaryIpToLowerdMap(secondary_ip_to_lowerd_elemen_nodal_map);
+          mc->setPrimaryIpToLowerdMap(primary_ip_to_lowerd_elemen_nodal_map);
+          mc->setNodalTangents(nodal_tangents);
```
This is my design so I'm not going to request that you change it, but seeing all these setters and getters used in this loop class makes me think that we should just have a reference to `AutomaticMortarGeneration` in the mortar constraint class. Then we could cut out the loop middle-man. Anyway that could be a follow-on cleanup

recuero replied:
I'll take a stab at this in a new commit.

lindsayad commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -634,6 +723,15 @@ ContactAction::addMortarContact()
+      // Additional displacement residual for frictional problem
+      // The second frictional LM acts on a perpendicular direction.
+      if (is_additional_frictional_constraint)
+      {
+        MooseEnum direction("direction_1 direction_2", "direction_2");
+        params.set<MooseEnum>("direction") = direction;
+      }
```
```suggestion
        params.set<MooseEnum>("direction") = "direction_2";
```
I'm pretty sure this should work

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalPressureVectorAux.C:
```diff
@@ -0,0 +1,124 @@
+               "MortarFrictionalPressureVector auxiliary kernel must be restricted to a boundary.");
+
+  // Get mortar interfaces
+  const auto & displaced_mortar_interfaces = _fe_problem.getMortarInterfaces(/*displaced=*/true);
```
Hmm could we set the `use_displaced_mesh` parameter to true in `validParams` and then do a `paramError` if it's false in the constructor?

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalPressureVectorAux.C:
```diff
@@ -0,0 +1,124 @@
+Real
+MortarFrictionalPressureVectorAux::computeValue()
+{
+  const auto & displaced_mortar_interfaces = _fe_problem.getMortarInterfaces(/*displaced=*/true);
+
+  // Get automatic generation object for the boundary pair this auxiliary acts on.
+  if (displaced_mortar_interfaces.count(std::make_pair(_primary_id, _secondary_id)) != 1)
+    mooseError("primary_boundary",
+               "The boundary pairs do not correspond to a single mortar contact boundary pair. "
+               "Please revise your input file for proper mortar contact constraints and mortar "
+               "frictional pressure vector auxiliary variable definition.");
+
+  const AutomaticMortarGeneration & mortar_generation_object =
+      displaced_mortar_interfaces.at(std::make_pair(_primary_id, _secondary_id));
```
Is it possible to do this setup in the constructor? It looks like you are already doing most of it there ... just have to create a pointer member for the automatic mortar generation object. Also let's use `libmesh_map_find` instead of `at` when we are confident a failed find would represent a developer failure. And then if a failed find would ever be due to a user input failure we should be doing an explicit map find and then `mooseError` (which I believe you are already doing in the constructor so 👍 )

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalPressureVectorAux.C:
```diff
@@ -0,0 +1,124 @@
+
+  Real tangent_one_component = 0;
+  Real tangent_two_component = 0;
+  //
```
```suggestion
```

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -138,6 +144,64 @@ AutomaticMortarGeneration::getNodalNormals(const Elem & secondary_elem) const
+
+  for (const auto n : make_range(secondary_elem.n_nodes()))
+  {
+    nodal_tangents_one[n] = _secondary_node_to_hh_nodal_tangents.at(secondary_elem.node_ptr(n))[0];
+    nodal_tangents_two[n] = _secondary_node_to_hh_nodal_tangents.at(secondary_elem.node_ptr(n))[1];
```
Use `libmesh_map_find` here

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalPressureVectorAux.C:
```diff
@@ -0,0 +1,124 @@
+  Real tangent_one_component = 0;
+  Real tangent_two_component = 0;
+  //
+  for (const auto lowerd_node : make_range(lower_dimensional_element->n_nodes()))
+  {
+    if (_fe_problem.mesh().getMesh().node_ptr(_current_node->id()) ==
+        _fe_problem.mesh().getMesh().node_ptr(lower_dimensional_element->node_id(lowerd_node)))
+    {
+      tangent_one_component = nodal_tangents[0][0](_component);
+      tangent_two_component = nodal_tangents[1][0](_component);
+      break;
+    }
+  }
```
I don't really understand what's going on here. I feel like the second index of `nodal_tangents` should be with `lowerd_node` as opposed to always 0?

lindsayad replied:
I don't think we can be sure that `element_ids[0]` zeroth node will always be the node that corresponds to `_current_node`

recuero replied:
> I don't really understand what's going on here. I feel like the second index of nodal_tangents should be with lowerd_node as opposed to always 0?

Yeah, you are right. I was testing this on a flat surface, so couldn't see the error there. Initially, I didn't have the logic and hard coded it to zero. When I put the logic together I forgot to actually use it.  I believe this update should also address your second comment.

lindsayad commented on modules/contact/src/constraints/ADMortarLagrangeConstraint.C:
```diff
@@ -0,0 +1,251 @@
+  }
+
+  for (_qp = 0; _qp < _qrule_msm->n_points(); _qp++)
+    for (const auto index : is_index_on_lower_dimension)
+    {
+      _i = index;
+      _local_re(_i) += raw_value(_JxW_msm[_qp] * _coord[_qp] * computeQpResidual(mortar_type));
+    }
```
I like it!!

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -166,9 +212,15 @@ ComputeFrictionalForceLMMechanicalContact::post()
+    _tangential_vel_ptr[0] = &(pr.second[0]);
 
-    enforceConstraintOnDof(dof);
+    if (_3d)
+    {
+      _tangential_vel_ptr[1] = &(pr.second[1]);
+      enforceConstraintOnDof3d(dof);
+    }
+    else
+      enforceConstraintOnDof(dof);
```
I appreciate all the simplification you've done

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -190,9 +242,93 @@ ComputeFrictionalForceLMMechanicalContact::incorrectEdgeDroppingPost(
+  }
+  else
+  {
+    Real epsilon_ad = 1.0e-14;
```
Maybe rename to `epsilon_sqrt`? I think the `ad` bit isn't as clear since its type is actually `Real`. Also make `const`

lindsayad commented on modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:
```diff
@@ -108,7 +107,12 @@ ComputeWeightedGapLMMechanicalContact::computeQpProperties()
         (*_primary_disp_z)[_qp].derivatives() - (*_secondary_disp_z)[_qp].derivatives();
 
-  _qp_gap = gap_vec * (_normals[_normal_index] * _JxW_msm[_qp] * _coord[_qp]);
+  if (_interpolate_normals)
+    _qp_gap = gap_vec * (_normals[_qp] * _JxW_msm[_qp] * _coord[_qp]);
+  else
+    _qp_gap_nodal = gap_vec * (_JxW_msm[_qp] * _coord[_qp]);
```
It's tough sometimes when I don't write the code to have an intuitive grasp for everything. For instance I don't know whether it makes sense to have two different gap members?

recuero replied:
To use the design of these mortar contact constraint, I think we need to make some distinction between the quantities from  fully interpolated geometry and those resulting from the nodal geometry. Here, I am trying to have a nodal vector of interpolated quantities (`_qp_gap_nodal`)  which will be paired with a nodal normal vector in `computeQpIProperties` when `_interpolate_normals` is false, whereas the `_qp_gap` is a scalar from the fully interpolated strategy for the default `_interpolate_normals` true.

lindsayad replied:
that sounds fine. Just wanted to pick your brain about it

lindsayad reviewed (changes requested):
Looks like no regolding necessary here?

recuero committed:
Address Alex's comments (#17495)

 - Also add markdown documentation for mortar frictional auxiliary kernel

recuero committed:
Have a reference to AutomaticMortarGeneration in the mortar constraint class. (#17495)

This slight design change suggested by Alex enables the mortar constraints to directly
interrogate mortar generation for geometry information, instead of preemtively
set required quantities from the mortar functor. So the mortar functor is cut out
from this transaction.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/921306/) on 4f3c85b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19462/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19462/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c65ead690c99f7dd907692b246bf4af004149018`

lindsayad reviewed (approved):
if we create a derived object from `AutomaticMortarGeneration` specific to contact, then we should move the methods that are unused in framework testing to that object
